_src/_posts/2019-02-26-year-of-hell.markdown

```diff
@@ -0,0 +1,454 @@
+and I knew the next N months of my life would be... well... "fun".
+
+
+## The project
+
+The project itself wasn't a rewrite, but the changes involved were huge.
+As I said, it all started with a new UI. The new UI started adding 
+constraints to the backend, enough that we ended up rewriting most of
+the existing code base. So it wasn't a rewrite per se, but we rewrote
+most of our code anyway.
+
+I mentioned we switched from a highly configurable fixed function pipeline
+into a highly flexible rules-based one that had a lot more expressive power.
+That was the core of the project.
+
+But that's easier said than done. The new system used rules instead
+of configuration. Each campaign would basically have a program that had
+to run in order to figure out whether to bid on an impression or not.
+
+Suddenly, all the optimizations we had in place to meet our CPU budget
+no longer applied, and we had to work out another way to optimize things.
+
+So... how do we get there?
+
+We had to solve quite a few nontrivial - nay, complex - problems in the
+process:
+
+* **The Bidder Decision Tree** (BDT)
+
+  First, we replaced all the straightforward business logic code by a
+  decision-tree-like structure that could quickly cull rules based on
+  auction characteristics.
+
+* **Instrumentation**
+
+  Since the BDT was so magical that it would be an inscrutable black box, and
+  that wasn't ok, we designed a whole new instrumentation
+  system to measure all the decisions taken within the BDT with minimal overhead.
+
+* **The new UI itself**
+
+  The new UI was indeed a complete rewrite from scratch, with microservices,
+  Node.js and React. It was a massive undertaking.
+
+* **Data migration**
+
+  The new model was radically different from the old model so, at some point, we had to
+  transition databases, applications, everything to it. It was
+  no small feat, especially since it had to be done without impacting day-to-day operations. It's not
+  like we could start the new system with a blank database and tell everybody to set
+  everything up again, manually, from scratch. We had to come up with an unobtrusive
+  transition plan.
+
+As you can probably tell from the preface, though, the project was as stressing as it was
+cool. Each step of the way encountered unexpected roadblocks, and was cause
+for many a flamewar.
+
+Where to begin...
+
+### The Bidder Decision Tree
+
+Instead of processing campaigns one by one, caching results as we did
+before, we precomputed a decision tree optimized for whatever the current
+campaign setup is at the time. A greedy algorithm would be free to
+shape that BDT the way it best suited the current setup, making it a very
+flexible design.
+
+The decision tree would trade offline CPU time with online CPU time.
+Building the tree would take a lot of CPU time, but the structure can
+then be shared by all bidders and that cost gets amortized in time and scale.
```

Comment:
`bidder instances` (or something like that).

Reply:
I don't think it adds much to outsiders. It may even be confusing for them.